src/pypi2nix/dependency_graph.py

```diff
@@ -0,0 +1,217 @@
+from collections import defaultdict
+from copy import copy
+from typing import Callable
+from typing import DefaultDict
+from typing import Dict
+from typing import Generator
+from typing import List
+from typing import Set
+from typing import TypeVar
+
+import yaml
+
+from pypi2nix.external_dependencies import ExternalDependency
+from pypi2nix.requirements import Requirement
+
+K = TypeVar("K")
+V = TypeVar("V")
+
+
+class DependencyGraph:
+    def __init__(self) -> None:
+        self._runtime_dependencies: DefaultDict[str, Set[str]] = defaultdict(
+            lambda: set()
+        )
+        self._buildtime_dependencies: DefaultDict[str, Set[str]] = defaultdict(
+            lambda: set()
+        )
+        self._external_dependencies: DefaultDict[
+            str, Set[ExternalDependency]
+        ] = defaultdict(lambda: set())
+
+    def set_runtime_dependency(
+        self, dependent: Requirement, dependency: Requirement
+    ) -> None:
+        if self.is_buildtime_dependency(dependency, dependent):
```

Comment:
This could and should be refactored into a function called `self.raise_on_cyclic_dependency()` or something similar.